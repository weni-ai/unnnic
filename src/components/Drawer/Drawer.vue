<template>
  <aside
    v-if="modelValue"
    class="unnnic-drawer"
  >
    <section
      v-if="!withoutOverlay"
      class="unnnic-drawer__overlay"
      @click.stop="close"
    />
    <Transition
      appear
      name="drawer"
    >
      <section
        v-if="showDrawer"
        :class="[
          'unnnic-drawer__container',
          wide && 'unnnic-drawer__container--wide',
        ]"
      >
        <header class="unnnic-drawer__header">
          <section class="unnnic-drawer__title-container">
            <h1 class="unnnic-drawer__title">{{ title }}</h1>
            <p
              v-if="description"
              class="unnnic-drawer__description"
            >
              {{ description }}
            </p>
          </section>
          <UnnnicIcon
            class="unnnic-drawer__close"
            :icon="closeIcon"
            size="avatar-nano"
            clickable
            @click="back"
          />
        </header>
        <section class="unnnic-drawer__content">
          <slot name="content"></slot>
        </section>
        <footer
          v-if="showFooter"
          class="unnnic-drawer__footer"
        >
          <UnnnicButton
            v-if="secondaryButtonText"
            size="large"
            type="tertiary"
            :disabled="disabledSecondaryButton"
            :loading="loadingSecondaryButton"
            :text="secondaryButtonText"
            @click="$emit('secondaryButtonClick')"
          />
          <UnnnicButton
            v-if="primaryButtonText"
            size="large"
            :disabled="disabledPrimaryButton"
            :loading="loadingPrimaryButton"
            :type="primaryButtonType"
            :text="primaryButtonText"
            @click="$emit('primaryButtonClick')"
          />
        </footer>
      </section>
    </Transition>
  </aside>
</template>

<script>
import UnnnicIcon from '../Icon.vue';
import UnnnicButton from '../Button/Button.vue';

export default {
  name: 'UnnnicDrawer',
  components: {
    UnnnicIcon,
    UnnnicButton,
  },
  props: {
    title: {
      type: String,
      required: true,
    },
    description: {
      type: String,
      default: '',
    },
    disabledPrimaryButton: {
      type: Boolean,
      default: false,
    },
    disabledSecondaryButton: {
      type: Boolean,
      default: false,
    },
    loadingPrimaryButton: {
      type: Boolean,
      default: false,
    },
    loadingSecondaryButton: {
      type: Boolean,
      default: false,
    },
    primaryButtonText: {
      type: String,
      default: '',
    },
    primaryButtonType: {
      type: String,
      default: 'primary',
    },
    secondaryButtonText: {
      type: String,
      default: '',
    },
    wide: {
      type: Boolean,
      default: false,
    },
    modelValue: {
      type: Boolean,
      required: true,
    },
    withoutOverlay: {
      type: Boolean,
      default: false,
    },
    closeIcon: {
      type: String,
      default: 'arrow_back',
    },
    distinctCloseBack: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['primaryButtonClick', 'secondaryButtonClick', 'close', 'back'],
  data() {
    return {
      showDrawer: true,
    };
  },
  computed: {
    showFooter() {
      return !!(this.primaryButtonText || this.secondaryButtonText);
    },
  },
  methods: {
    transitionClose(callback) {
      this.showDrawer = false;
      setTimeout(() => {
        callback?.();
        this.showDrawer = true;
      }, 200);
    },
    close() {
      this.transitionClose(() => this.$emit('close'));
    },
    back() {
      if (this.distinctCloseBack) {
        this.transitionClose(() => this.$emit('back'));
      } else {
        this.close();
      }
    },
  },
};
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

@keyframes drawerOpen {
  from {
    transform: translateX(100%);
  }
  to {
    transform: translateX(0);
  }
}

@keyframes drawerClose {
  from {
    transform: translateX(0);
  }
  to {
    transform: translateX(100%);
  }
}

.drawer-enter-active,
.drawer-enter-to {
  animation: drawerOpen 200ms ease-in;
}

.drawer-leave-active,
.drawer-leave-to {
  display: none;
  animation: drawerClose 200ms ease-in;
}

.unnnic-drawer {
  width: 100vw;
  height: 100vh;
  position: fixed;
  left: 0;
  top: 0;
  z-index: 8;
}

.unnnic-drawer__overlay {
  z-index: 9;
  background-color: rgba(0, 0, 0, 0.4);
  width: 100%;
  height: 100%;
}

.unnnic-drawer__container {
  z-index: 10;
  top: 0;
  right: 0;
  position: fixed;
  display: flex;
  flex-direction: column;
  font-family: $unnnic-font-family-secondary;
  justify-content: space-between;
  height: 100%;
  background-color: $unnnic-color-neutral-white;
  width: calc(100% / 3);

  &--wide {
    width: 50%;
  }

  .unnnic-drawer__header {
    display: flex;
    justify-content: space-between;
    border-bottom: 1px solid $unnnic-color-neutral-soft;
    padding: $unnnic-spacing-md;
    .unnnic-drawer__title-container {
      .unnnic-drawer__title {
        color: $unnnic-color-neutral-darkest;
        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-title-sm;
        font-weight: $unnnic-font-weight-black;
        line-height: $unnnic-line-height-large * 1.75;
      }

      .unnnic-drawer__description {
        color: $unnnic-color-neutral-cloudy;
        font-family: $unnnic-font-family-secondary;
        font-size: $unnnic-font-size-body-gt;
        font-weight: $unnnic-font-weight-regular;
        line-height: $unnnic-line-height-large * 1.375;
      }
    }

    .unnnic-drawer__close {
      margin: $unnnic-spacing-nano;
      transform: rotate(180deg);
      display: flex;
    }
  }

  .unnnic-drawer__content {
    overflow-y: auto;
    color: $unnnic-color-neutral-cloudy;
    padding: $unnnic-spacing-md $unnnic-spacing-md 0 $unnnic-spacing-md;
    flex: 1 0 0;
  }

  .unnnic-drawer__footer {
    display: flex;
    padding: $unnnic-spacing-md;
    gap: $unnnic-spacing-ant;
    > * {
      flex-grow: 1;
    }
  }
}
</style>
