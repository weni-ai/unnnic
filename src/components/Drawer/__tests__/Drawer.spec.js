import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach } from 'vitest';
import Drawer from '../Drawer.vue';

describe('Drawer.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(Drawer, {
      props: {
        modelValue: true,
        title: 'Test Title',
        description: 'Test Description',
        primaryButtonText: 'Primary Action',
        secondaryButtonText: 'Secondary Action',
      },
      global: {
        stubs: ['UnnnicIcon', 'UnnnicButton'],
      },
    });
  });

  describe('Component Rendering', () => {
    it('should render the drawer when modelValue is true', () => {
      expect(wrapper.find('[data-testid="drawer"]').exists()).toBe(true);
    });

    it('should not render the drawer when modelValue is false', async () => {
      await wrapper.setProps({ modelValue: false });
      expect(wrapper.find('[data-testid="drawer"]').exists()).toBe(false);
    });

    it('should render the overlay when withoutOverlay is false', () => {
      expect(wrapper.find('[data-testid="overlay"]').exists()).toBe(true);
    });

    it('should not render the overlay when withoutOverlay is true', async () => {
      await wrapper.setProps({ withoutOverlay: true });
      expect(wrapper.find('[data-testid="overlay"]').exists()).toBe(false);
    });

    it('should display the title and description correctly', () => {
      expect(wrapper.find('[data-testid="drawer-title"]').text()).toBe(
        'Test Title',
      );
      expect(wrapper.find('[data-testid="drawer-description"]').text()).toBe(
        'Test Description',
      );
    });

    it('should not display the description when not provided', async () => {
      await wrapper.setProps({ description: '' });
      expect(wrapper.find('[data-testid="drawer-description"]').exists()).toBe(
        false,
      );
    });

    it('should render the slot content correctly', () => {
      wrapper = mount(Drawer, {
        props: {
          modelValue: true,
          title: 'Test Title',
        },
        slots: {
          content: '<div data-testid="slot-content">Slot Content</div>',
        },
      });
      expect(wrapper.find('[data-testid="slot-content"]').exists()).toBe(true);
      expect(wrapper.find('[data-testid="slot-content"]').text()).toBe(
        'Slot Content',
      );
    });
  });

  describe('Button Behavior', () => {
    it('should emit the primaryButtonClick event when the primary button is clicked', async () => {
      await wrapper
        .findComponent('[data-testid="primary-button"]')
        .trigger('click');
      expect(wrapper.emitted('primaryButtonClick')).toBeTruthy();
    });

    it('should disable the primary button when disabledPrimaryButton is true', async () => {
      await wrapper.setProps({ disabledPrimaryButton: true });
      expect(
        wrapper.find('[data-testid="primary-button"]').attributes('disabled'),
      ).toBeDefined();
    });

    it('should display the correct text on the primary button', () => {
      expect(
        wrapper.findComponent('[data-testid="primary-button"]').props('text'),
      ).toBe('Primary Action');
    });

    it('should emit the secondaryButtonClick event when the secondary button is clicked', async () => {
      await wrapper
        .findComponent('[data-testid="secondary-button"]')
        .trigger('click');
      expect(wrapper.emitted('secondaryButtonClick')).toBeTruthy();
    });

    it('should disable the secondary button when disabledSecondaryButton is true', async () => {
      await wrapper.setProps({ disabledSecondaryButton: true });
      expect(
        wrapper
          .findComponent('[data-testid="secondary-button"]')
          .attributes('disabled'),
      ).toBeDefined();
    });

    it('should display the correct text on the secondary button', () => {
      expect(
        wrapper.findComponent('[data-testid="secondary-button"]').props('text'),
      ).toBe('Secondary Action');
    });

    it('should render the footer only if showFooter is true', async () => {
      expect(wrapper.find('[data-testid="footer"]').exists()).toBe(true);

      await wrapper.setProps({
        primaryButtonText: '',
        secondaryButtonText: '',
      });

      expect(wrapper.find('[data-testid="footer"]').exists()).toBe(false);
    });
  });

  describe('Interactions and Transitions', () => {
    it('should close the drawer when the close icon is clicked', async () => {
      await wrapper
        .findComponent('[data-testid="close-icon"]')
        .trigger('click');

      setTimeout(() => {
        expect(wrapper.emitted('close')).toBeTruthy();
        expect(wrapper.vm.showDrawer).toBe(false);
      }, 200); // Transition time
    });

    it('should call back method when back icon is clicked and distinctCloseBack is true', async () => {
      await wrapper.setProps({ distinctCloseBack: true });
      await wrapper
        .findComponent('[data-testid="close-icon"]')
        .trigger('click');

      setTimeout(() => {
        expect(wrapper.emitted('back')).toBeTruthy();
      }, 200); // Transition time
    });

    it('should not emit back event when back icon is clicked and distinctCloseBack is false', async () => {
      await wrapper.find('[data-testid="close-icon"]').trigger('click');
      expect(wrapper.emitted('back')).toBeUndefined();
    });

    it('should correctly handle transitions when closing the drawer', async () => {
      await wrapper.vm.close();
      expect(wrapper.vm.showDrawer).toBe(false);

      setTimeout(() => {
        expect(wrapper.find('[data-testid="drawer"]').exists()).toBe(false);
      }, 200); // Transition time
    });

    it('should handle overlay clicks to close the drawer when withoutOverlay is false', async () => {
      expect(wrapper.find('[data-testid="overlay"]').exists()).toBe(true);
      await wrapper.find('[data-testid="overlay"]').trigger('click');
      setTimeout(() => {
        expect(wrapper.emitted('close')).toBeTruthy();
      }, 200); // Transition time
    });
  });

  describe('Props and Computed Properties', () => {
    it('should render the correct size class based on size prop', async () => {
      await wrapper.setProps({ size: 'lg' });
      expect(
        wrapper.find('[data-testid="drawer-container"]').classes(),
      ).toContain('unnnic-drawer__container--lg');

      await wrapper.setProps({ size: 'xl' });
      expect(
        wrapper.find('[data-testid="drawer-container"]').classes(),
      ).toContain('unnnic-drawer__container--xl');
    });

    it('should display footer if either primaryButtonText or secondaryButtonText is provided', async () => {
      expect(wrapper.find('[data-testid="footer"]').exists()).toBe(true);

      await wrapper.setProps({
        primaryButtonText: '',
        secondaryButtonText: 'Secondary Button',
      });
      expect(wrapper.find('[data-testid="footer"]').exists()).toBe(true);
    });

    it('should not display footer if both primaryButtonText and secondaryButtonText are not provided', async () => {
      await wrapper.setProps({
        primaryButtonText: '',
        secondaryButtonText: '',
      });

      expect(wrapper.find('[data-testid="footer"]').exists()).toBe(false);
    });

    it('should disable primary button when disabledPrimaryButton prop is true', async () => {
      await wrapper.setProps({
        primaryButtonText: 'Primary Button',
        disabledPrimaryButton: true,
      });
      expect(
        wrapper.find('[data-testid="primary-button"]').attributes('disabled'),
      ).toBe('true');
    });

    it('should disable secondary button when disabledSecondaryButton prop is true', async () => {
      await wrapper.setProps({
        secondaryButtonText: 'Secondary Button',
        disabledSecondaryButton: true,
      });
      expect(
        wrapper.find('[data-testid="secondary-button"]').attributes('disabled'),
      ).toBe('true');
    });

    it('should use correct icon for close button based on closeIcon prop', async () => {
      await wrapper.setProps({ closeIcon: 'custom_close_icon' });
      const closeIcon = wrapper.findComponent('[data-testid="close-icon"]');
      expect(closeIcon.props('icon')).toBe('custom_close_icon');
    });

    it('should validate size prop to accept only md, lg, or xl values', async () => {
      const validator = Drawer.props.size.validator;
      expect(validator('md')).toBe(true);
      expect(validator('lg')).toBe(true);
      expect(validator('xl')).toBe(true);
      expect(validator('sm')).toBe(false); // Should not be valid
    });
  });
});
