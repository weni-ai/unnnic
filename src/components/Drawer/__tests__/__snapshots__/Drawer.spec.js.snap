// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Drawer.vue > Component Rendering > Component Rendering > matches the snapshot 1`] = `
"<aside data-v-196de012="" class="unnnic-drawer" data-testid="drawer">
  <section data-v-196de012="" class="unnnic-drawer__overlay" data-testid="overlay"></section>
  <transition-stub data-v-196de012="" name="drawer" appear="true" persisted="false" css="true">
    <section data-v-196de012="" data-testid="drawer-container" class="unnnic-drawer__container unnnic-drawer__container--md">
      <header data-v-196de012="" class="unnnic-drawer__header">
        <section data-v-196de012="" class="unnnic-drawer__title-container">
          <h1 data-v-196de012="" class="unnnic-drawer__title" data-testid="drawer-title">Test Title</h1>
          <p data-v-196de012="" class="unnnic-drawer__description" data-testid="drawer-description">Test Description</p>
        </section>
        <unnnic-icon-stub data-v-196de012="" filled="false" icon="arrow_back" clickable="true" size="avatar-nano" scheme="neutral-darkest" class="unnnic-drawer__close" data-testid="close-icon"></unnnic-icon-stub>
      </header>
      <section data-v-196de012="" class="unnnic-drawer__content"></section>
      <footer data-v-196de012="" class="unnnic-drawer__footer" data-testid="footer">
        <unnnic-button-stub data-v-196de012="" size="large" text="Secondary Action" type="tertiary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" disabled="false" loading="false" data-testid="secondary-button"></unnnic-button-stub>
        <unnnic-button-stub data-v-196de012="" size="large" text="Primary Action" type="primary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" disabled="false" loading="false" data-testid="primary-button"></unnnic-button-stub>
      </footer>
    </section>
  </transition-stub>
</aside>"
`;
